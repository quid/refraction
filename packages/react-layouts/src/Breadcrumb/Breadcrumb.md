Breadcrumb example:
It allows users to keep track and maintain awareness of their locations within your website.

```js
const { MemoryRouter } = require('react-router-dom');
const items = [
  {
    label: 'Project Name 13',
    path: '/',
    arrowIcon: 'angle_double_right',
    emphasized: true,
  },
  {
    label: 'SEARCH',
    path: '/search',
    tooltip: 'project AND name',
  },
  {
    label: 'EXPLORE',
    path: '/search/explore',
  },
  {
    label: 'ANALYZE',
    path: '/Breadcrumb',
    disabled: true,
  },
  {
    label: 'PRESENT',
    path: '/search/explore/analyze/present',
    disabled: true,
  },
  {
    label: 'EXTERNAL',
    path: 'https://quid.com',
    external: true,
  },
  {
    label: 'Quid custom',
    path: 'https://quid.com',
    external: true,
    renderContent: ({ label, path, index, external }) => (
      <div>
        <a href={path}>
          {label} {index}
        </a>
      </div>
    ),
  },
];

<MemoryRouter>
  <Breadcrumb items={items} />
</MemoryRouter>;
```
