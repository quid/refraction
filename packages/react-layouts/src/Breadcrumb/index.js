/**
 * Copyright (c) Quid, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
// @flow
import * as React from 'react';
import NavLink from './NavLink';
import { Icon } from '@quid/react-core';
import styled from '@emotion/styled/macro';
import { textStyles } from '@quid/theme';
import { sizes } from '@quid/theme';

type BreadcrumbItem = {
  label: React.Node,
  path: string,
  arrowIcon?: string,
  disabled?: boolean,
  external?: boolean,
  emphasized?: boolean,
  tooltip?: string,
};

export type Props = {
  items: Array<BreadcrumbItem>,
  className?: string,
  separator?: React.Node,
};

const Arrow = styled.div`
  padding-left: ${sizes.regular};
  padding-right: ${sizes.regular};
  ${textStyles('normal', 'disabled')};
`;

type ItemType = BreadcrumbItem & {
  index: number,
  component?: React.Node,
};

const Item = styled(
  ({
    index,
    label,
    path,
    disabled = false,
    external = false,
    emphasized = false,
    tooltip = '',
    ...props
  }: ItemType) => (
    <NavLink
      to={path}
      emphasized={emphasized}
      external={external}
      disabled={disabled}
      key={index}
      {...props}
    >
      {tooltip ? (
        <span title={tooltip} id={`breadcrumb-${index}-tooltip`}>
          {label}
        </span>
      ) : (
        label
      )}
    </NavLink>
  )
)``;

const Breadcrumb = styled(({ items, separator, ...props }: Props) => (
  <nav {...props}>
    {items.reduce((accumulator, accItem, index) => {
      const { arrowIcon = 'angle_right', ...item } = accItem;
      const isLast = items.length === index + 1;
      accumulator.push(<Item key={`${index}_item`} index={index} {...item} />);

      if (isLast === false) {
        if (separator) {
          accumulator.push(
            <React.Fragment key={`${index}_arrow`}>{separator}</React.Fragment>
          );
        } else {
          accumulator.push(
            <Arrow key={`${index}_arrow`}>
              <Icon name={arrowIcon} />
            </Arrow>
          );
        }
      }
      return accumulator;
    }, [])}
  </nav>
))`
  display: flex;
  padding: 0;
  align-items: baseline;
`;

Breadcrumb.Item = Item;

// @component
export default Breadcrumb;
