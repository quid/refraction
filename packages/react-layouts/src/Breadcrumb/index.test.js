/**
 * Copyright (c) Quid, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

// @flow
import React from 'react';
import { mount } from 'enzyme';
import Breadcrumb from './index';
import NavLink from './NavLink';
import { BrowserRouter } from 'react-router-dom';

const items = [
  {
    label: 'Project Name 13',
    path: '/',
    arrowIcon: 'angle_double_right',
  },
  {
    label: 'SEARCH',
    path: '/search',
    tooltip: 'project AND name',
  },
  {
    label: 'REVIEW',
    path: '/search/review',
  },
  {
    label: 'ANALYZE',
    path: '/search/review/analyze',
    disabled: true,
  },
  {
    label: 'SHARE',
    path: '/search/review/analyze/share',
    external: true,
    emphasized: true,
  },
];

it('renders the expected markup', () => {
  const wrapper = mount(
    <BrowserRouter>
      <Breadcrumb items={items} />
    </BrowserRouter>
  );

  expect(wrapper).toMatchSnapshot();
});

it('renders a default NavLink', () => {
  const wrapper = mount(
    <BrowserRouter>
      <NavLink to="foobar" />
    </BrowserRouter>
  );

  expect(wrapper).toMatchSnapshot();
});

it('renders a NavLink', () => {
  const wrapper = mount(
    <BrowserRouter>
      <NavLink to="foobar" external disabled activeClassName="foobar--active" />
    </BrowserRouter>
  );

  expect(wrapper).toMatchSnapshot();
});

it('renders x amount of Breadcrumb separators', () => {
  const wrapper = mount(
    <BrowserRouter>
      <Breadcrumb items={items} separator={<div data-context="separator" />} />
    </BrowserRouter>
  );

  expect(wrapper.find('[data-context="separator"]').length).toBe(
    items.length - 1
  );

  expect(wrapper.find('[data-context="separator"]').first()).toMatchSnapshot();
});

it('renders an item where label as defined', () => {
  const itemsWithChildren = [
    ...items,
    {
      label: <div data-context="label">Children</div>,

      path: '/children',
    },
  ];
  const wrapper = mount(
    <BrowserRouter>
      <Breadcrumb items={itemsWithChildren} />
    </BrowserRouter>
  );

  expect(wrapper.find('[data-context="label"]').first()).toMatchSnapshot();
});
