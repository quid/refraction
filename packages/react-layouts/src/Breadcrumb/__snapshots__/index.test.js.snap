// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a NavLink 1`] = `
.emotion-0 {
  font-family: IBM Plex Sans,Lucida Grande,Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 1.57;
  color: #2E3338;
  font-size: 14px;
  line-height: 1.57;
  color: #2E3338;
  color: #2E3338;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: IBM Plex Sans,Lucida Grande,Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 1.57;
  color: #2E3338;
  color: #8F9BA3;
  cursor: default;
}

<NavLink
  activeClassName="foobar--active"
  disabled={true}
  external={true}
  to="foobar"
>
  <a
    activeClassName="foobar--active"
    className="emotion-0 emotion-1"
  />
</NavLink>
`;

exports[`renders a default NavLink 1`] = `
.emotion-0 {
  font-family: IBM Plex Sans,Lucida Grande,Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 1.57;
  color: #2E3338;
  font-size: 14px;
  line-height: 1.57;
  color: #2E3338;
  color: #2E3338;
  -webkit-text-decoration: none;
  text-decoration: none;
}

<NavLink
  to="foobar"
>
  <ForwardRef(render)>
    <a
      aria-current={null}
      className="emotion-0 emotion-1"
      href="/foobar"
      onClick={[Function]}
    />
  </ForwardRef(render)>
</NavLink>
`;

exports[`renders an item where label as defined 1`] = `
<div
  data-context="label"
>
  Children
</div>
`;

exports[`renders the expected markup 1`] = `
.emotion-42 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
}

.emotion-1 {
  font-family: IBM Plex Sans,Lucida Grande,Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 1.57;
  color: #2E3338;
  font-size: 14px;
  line-height: 1.57;
  color: #2E3338;
  color: #2E3338;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-3 {
  font-family: IBM Plex Sans,Lucida Grande,Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 1.57;
  color: #2E3338;
  font-weight: bold;
}

.emotion-8 {
  padding-left: 10px;
  padding-right: 10px;
  font-family: IBM Plex Sans,Lucida Grande,Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 1.57;
  color: #2E3338;
  font-size: 14px;
  line-height: 1.57;
  color: #2E3338;
  color: #8F9BA3;
}

.emotion-6 {
  font-family: 'quid-icons' !important;
  speak: none;
  font-style: normal;
  font-weight: normal;
  font-variant: normal;
  text-transform: none;
  line-height: inherit;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  font-feature-settings: 'liga';
  font-variant-ligatures: discretionary-ligatures;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  height: 1em;
  width: 1em;
}

.emotion-29 {
  font-family: IBM Plex Sans,Lucida Grande,Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 1.57;
  color: #2E3338;
  font-size: 14px;
  line-height: 1.57;
  color: #2E3338;
  color: #2E3338;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: IBM Plex Sans,Lucida Grande,Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 1.57;
  color: #2E3338;
  color: #8F9BA3;
  cursor: default;
}

.emotion-38 {
  font-family: IBM Plex Sans,Lucida Grande,Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 1.57;
  color: #2E3338;
  font-size: 14px;
  line-height: 1.57;
  color: #2E3338;
  color: #2E3338;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: IBM Plex Sans,Lucida Grande,Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 1.57;
  color: #2E3338;
  font-weight: bold;
}

<Breadcrumb
  items={
    Array [
      Object {
        "arrowIcon": "angle_double_right",
        "label": "Project Name 13",
        "path": "/",
      },
      Object {
        "label": "SEARCH",
        "path": "/search",
        "tooltip": "project AND name",
      },
      Object {
        "label": "REVIEW",
        "path": "/search/review",
      },
      Object {
        "disabled": true,
        "label": "ANALYZE",
        "path": "/search/review/analyze",
      },
      Object {
        "emphasized": true,
        "external": true,
        "label": "SHARE",
        "path": "/search/review/analyze/share",
      },
    ]
  }
>
  <nav
    className="emotion-42 emotion-43"
  >
    <Item
      index={0}
      key="0_item"
      label="Project Name 13"
      path="/"
    >
      <NavLink
        className="emotion-4 emotion-0"
        disabled={false}
        emphasized={false}
        external={false}
        key="0"
        to="/"
      >
        <ForwardRef(render)>
          <a
            aria-current="page"
            className="emotion-0 emotion-1 emotion-2 emotion-3"
            href="/"
            onClick={[Function]}
            style={Object {}}
          >
            Project Name 13
          </a>
        </ForwardRef(render)>
      </NavLink>
    </Item>
    <Arrow
      key="0_arrow"
    >
      <div
        className="emotion-8 emotion-9"
      >
        <Icon
          name="angle_double_right"
          squared={true}
        >
          <i
            className="emotion-6 emotion-7"
          >
            angle_double_right
          </i>
        </Icon>
      </div>
    </Arrow>
    <Item
      index={1}
      key="1_item"
      label="SEARCH"
      path="/search"
      tooltip="project AND name"
    >
      <NavLink
        className="emotion-4 emotion-0"
        disabled={false}
        emphasized={false}
        external={false}
        key="1"
        to="/search"
      >
        <ForwardRef(render)>
          <a
            aria-current={null}
            className="emotion-0 emotion-1 emotion-2"
            href="/search"
            onClick={[Function]}
          >
            <span
              id="breadcrumb-1-tooltip"
              title="project AND name"
            >
              SEARCH
            </span>
          </a>
        </ForwardRef(render)>
      </NavLink>
    </Item>
    <Arrow
      key="1_arrow"
    >
      <div
        className="emotion-8 emotion-9"
      >
        <Icon
          name="angle_right"
          squared={true}
        >
          <i
            className="emotion-6 emotion-7"
          >
            angle_right
          </i>
        </Icon>
      </div>
    </Arrow>
    <Item
      index={2}
      key="2_item"
      label="REVIEW"
      path="/search/review"
    >
      <NavLink
        className="emotion-4 emotion-0"
        disabled={false}
        emphasized={false}
        external={false}
        key="2"
        to="/search/review"
      >
        <ForwardRef(render)>
          <a
            aria-current={null}
            className="emotion-0 emotion-1 emotion-2"
            href="/search/review"
            onClick={[Function]}
          >
            REVIEW
          </a>
        </ForwardRef(render)>
      </NavLink>
    </Item>
    <Arrow
      key="2_arrow"
    >
      <div
        className="emotion-8 emotion-9"
      >
        <Icon
          name="angle_right"
          squared={true}
        >
          <i
            className="emotion-6 emotion-7"
          >
            angle_right
          </i>
        </Icon>
      </div>
    </Arrow>
    <Item
      disabled={true}
      index={3}
      key="3_item"
      label="ANALYZE"
      path="/search/review/analyze"
    >
      <NavLink
        className="emotion-4 emotion-0"
        disabled={true}
        emphasized={false}
        external={false}
        key="3"
        to="/search/review/analyze"
      >
        <a
          className="emotion-0 emotion-29 emotion-2"
        >
          ANALYZE
        </a>
      </NavLink>
    </Item>
    <Arrow
      key="3_arrow"
    >
      <div
        className="emotion-8 emotion-9"
      >
        <Icon
          name="angle_right"
          squared={true}
        >
          <i
            className="emotion-6 emotion-7"
          >
            angle_right
          </i>
        </Icon>
      </div>
    </Arrow>
    <Item
      emphasized={true}
      external={true}
      index={4}
      key="4_item"
      label="SHARE"
      path="/search/review/analyze/share"
    >
      <NavLink
        className="emotion-4 emotion-0"
        disabled={false}
        emphasized={true}
        external={true}
        key="4"
        to="/search/review/analyze/share"
      >
        <a
          className="emotion-0 emotion-38 emotion-2"
          href="/search/review/analyze/share"
        >
          SHARE
        </a>
      </NavLink>
    </Item>
  </nav>
</Breadcrumb>
`;

exports[`renders x amount of Breadcrumb separators 1`] = `
<div
  data-context="separator"
/>
`;
